File: development/NET/samples/server/AddressSpace/CS/MyTransaction.cs
using System;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace AddressSpace
{
	public class MyTransaction : DaTransaction
	{

		public MyTransaction(
			EnumTransactionType aTransactionType,
			DaRequest[] aRequests,
			uint aSessionKey) :
			base(aTransactionType, aRequests, aSessionKey)
		{

		}   //	end constructor

		public override int HandleReadRequests()
		{

			int count = Requests.Count;
			for (int i = 0; i < count; i++)
			{

				DaRequest request = Requests[i] as DaRequest;
				if ((request != null) && (request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) == true))
				{
					if (request.PropertyId == 0)
					{
						// get address space element value
						// take the toolkit cache value
						ValueQT valueQT = null;
						request.AddressSpaceElement.GetCacheValue(ref valueQT);
						request.Value = valueQT;
						request.Result = EnumResultCode.S_OK;
					}
					else
					{
						// get property value
						// get the value from the address space element
						MyDaAddressSpaceElement element = request.AddressSpaceElement as MyDaAddressSpaceElement;
						request.Value = new ValueQT(element.ObjectType, EnumQuality.GOOD, DateTime.Now);
						request.Result = EnumResultCode.S_OK;
					}   //	end if ... else
				}   //	end if
			}   //	end for

			return CompleteRequests();

		}   //	end HandleReadRequests

		public override int HandleWriteRequests()
		{

			int result = ValuesChanged();
			if (ResultCode.FAILED(result))
			{
				return result;
			}   //	end if

			return CompleteRequests();

		}   //	end HandleWriteRequests
	}   //	end class MyTransaction
}   //	end namespace

File: development/NET/samples/server/AddressSpace/VB/MyTransaction.vb
Imports System
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace AddressSpace

	Public Class MyTransaction
		Inherits DaTransaction

		Public Sub New(ByVal aTransactionType As EnumTransactionType, ByVal aRequests As DaRequest(), ByVal aSessionKey As UInt32)
			MyBase.New(aTransactionType, aRequests, aSessionKey)

		End Sub

		Public Overloads Overrides Function HandleReadRequests() As Integer

			Dim count As Integer = Requests.Count
			Dim i As Integer = 0

			While i < count
				Dim request As DaRequest = DirectCast(Requests(i), DaRequest)

				If request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) = True Then
					If request.PropertyId = 0 Then
						' get address space element value
						' take the toolkit cache value
						Dim valueQT As ValueQT = Nothing
						request.AddressSpaceElement.GetCacheValue(valueQT)
						request.Value = valueQT
						request.Result = EnumResultCode.S_OK
					Else
						' get property value
						' get the value from the address space element
						Dim element As MyDaAddressSpaceElement = DirectCast(request.AddressSpaceElement, MyDaAddressSpaceElement)
						Dim qualityGood As EnumQuality = CType([Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), EnumQuality)
						request.Value = New ValueQT(element.ObjectType, qualityGood, DateTime.Now)
						request.Result = EnumResultCode.S_OK
					End If
				End If

				System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
			End While

			Return CompleteRequests()

		End Function

		Public Overloads Overrides Function HandleWriteRequests() As Integer

			Dim result As Integer = ValuesChanged()
			If ResultCode.FAILED(result) Then
				Return result
			End If

			Return CompleteRequests()

		End Function
	End Class
End Namespace

